service/src/test/java/bio/terra/workspace/app/controller/WorkspaceApiControllerConnectedTest.java

```diff
@@ -443,6 +446,79 @@ public void listValidRegions() throws Exception {
     assertTrue(azures.isEmpty());
   }
 
+  @Test
+  @EnabledIf(expression = "${feature.tps-enabled}", loadContext = true)
+  public void updatePolicies_tpsEnabledAndPolicyUpdated() throws Exception {
+    ApiWorkspaceDescription workspaceWithoutPolicy =
+        mockMvcUtils.getWorkspace(userAccessUtils.defaultUserAuthRequest(), workspace.getId());
+    assertTrue(workspaceWithoutPolicy.getPolicies().isEmpty());
+
+    // add attributes
+    String usRegion = "US";
+    ApiWsmPolicyUpdateResult result =
+        mockMvcUtils.updateRegionPolicy(
+            userAccessUtils.defaultUserAuthRequest(), workspace.getId(), usRegion);
+
+    assertTrue(result.isUpdateApplied());
+    ApiWorkspaceDescription updatedWorkspace =
+        mockMvcUtils.getWorkspace(userAccessUtils.defaultUserAuthRequest(), workspace.getId());
+    assertEquals(1, updatedWorkspace.getPolicies().size());
+    assertEquals(
+        usRegion, updatedWorkspace.getPolicies().get(0).getAdditionalData().get(0).getValue());
+
+    // remove attributes
+    ApiWsmPolicyUpdateResult removeResult =
+        mockMvcUtils.removeRegionPolicy(
+            userAccessUtils.defaultUserAuthRequest(), workspace.getId(), usRegion);
+
+    assertTrue(removeResult.isUpdateApplied());
+    workspaceWithoutPolicy =
+        mockMvcUtils.getWorkspace(userAccessUtils.defaultUserAuthRequest(), workspace.getId());
+    assertEquals(0, workspaceWithoutPolicy.getPolicies().size());
+  }
+
+  @Test
+  @EnabledIf(expression = "${feature.tps-enabled}", loadContext = true)
+  public void updatePolicies_tpsEnabledAndPolicyConflict() throws Exception {
+    ApiWorkspaceDescription workspaceWithoutPolicy =
+        mockMvcUtils.getWorkspace(userAccessUtils.defaultUserAuthRequest(), workspace.getId());
+    assertTrue(workspaceWithoutPolicy.getPolicies().isEmpty());
+
+    var usRegion = "us";
+    ApiWsmPolicyUpdateResult result =
+        mockMvcUtils.updateRegionPolicy(
+            userAccessUtils.defaultUserAuthRequest(), workspace.getId(), usRegion);
+
+    assertTrue(result.isUpdateApplied());
+    ApiWorkspaceDescription updatedWorkspace =
+        mockMvcUtils.getWorkspace(userAccessUtils.defaultUserAuthRequest(), workspace.getId());
+    assertEquals(1, updatedWorkspace.getPolicies().size());
+
+    mockMvcUtils.updatePoliciesExpect(
+        userAccessUtils.defaultUserAuthRequest(),
+        workspace.getId(),
+        HttpStatus.SC_INTERNAL_SERVER_ERROR,
```

Comment:
I'm surprised this is a 500 error. Where is that coming from?
If this is a conflict, we should be returning success with conflict information.

So see, the test already has found a bug! :)

Not sure if we should apply the test until we get that straighted out. wdyt?

Reply:
oh ok. I wasn't sure what's the right behavior is.

the fix should be in TPS right? we are getting 409 from TPS and then wrap it in a 500 in TpsApidispatch. are you saying updatePao should return 200 for ENFORCE_CONFLICT and put conflict information in the response?

Reply:
409 are wrapped now.